crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -353,9 +358,59 @@ impl<TX: DbTx> DatabaseProvider<TX> {
         )
     }
 
-    /// Returns a reference to the [`ChainSpec`].
-    pub fn chain_spec(&self) -> &ChainSpec {
-        &self.chain_spec
+    fn block_with_senders_range<H: AsRef<Header>, B>(
+        &self,
+        headers_range: impl FnOnce(RangeInclusive<BlockNumber>) -> ProviderResult<Vec<H>>,
+        mut assemble_block: impl FnMut(
+            H,
+            Vec<TransactionSigned>,
+            Vec<Header>,
+            Option<Withdrawals>,
+            Option<Requests>,
+            Vec<Address>,
+        ) -> ProviderResult<B>,
+        range: RangeInclusive<BlockNumber>,
```

Comment:
we already have generics on the function, I'd like to do the same for these closures and not use impl